### ISSUE TITLE
Severe packet loss when using srt as a forwarding server.

### ISSUE THREAD

**Issue by mycrl** [Haivision/srt#2958]
I wrote a media server, a typical forwarding server model, one sender, multiple receivers, but I found that the packet loss from the sender to the server was particularly serious (at high bitrates).

* I didn't use srt group.
* I don't use non-blocking interfaces, all calls are blocking.
* The loss is serious under Windows, but I only tested Windows and Android. I think it is related to the high bit rate. This is because I use qsv to encode h264 on Windows. At present, it seems that there is no way to control the bit rate limit of qsv, so the data output on Windows is about 30Mbps, but I will deal with this problem later and reduce the bit rate.
* My programs all work in an intranet with a good network environment. Of course, I tested WIFI and network cable and found no difference.
* And excluding equipment performance issues, the network throughput of my equipment has reached at least one gigabit.
* In high bitrate situations, or when the bitrate suddenly increases, I observed from the network card data that the SRT data transmission is completely paused. This situation does not occur on the forwarding server, but on the Windows SRT sender. The forwarding server reports a large number of delay warning logs.
* It should be noted that the servers are all implemented in the same platform. I created SRT sockets from Windows and Android to connect to the same server. The bit rate on Windows is high, and the bit rate on Android is low. There is serious packet loss on Windows, but no problem on Android.

About forwarding server:

* I create a normal srt socket, and then use srt_bind and srt_listen to bind an address and port to use as a server.
* After that, I use srt_accept to accept the srt sockets from other clients. Of course, I will mark which srt socket is the sender and which srt socket is the receiver.
* I read the data from the sender and send it to all the receivers, repeating this process over and over again.

My options:

* SRTO_TRANSTYPE=SRTT_LIVE
* SRTO_RCVLATENCY=30
* SRTO_TSBPDMODE=true
* SRTO_RCVSYN=true
* SRTO_TLPKTDROP=true
* SRTO_REUSEADDR=true

Can't an ordinary SRT socket handle this duplex business?
I have observed that when the bitrate suddenly spikes, the srt packet loss also increases.  
or, Is there an incorrect implementation of SRT on Windows?

### srt version: 1.5.3

**Comment by mycrl** [comment]
Update: By setting SRTO_OHEADBW to 100%, the packet loss issue is mitigated, but I'd still like to know why

**Comment by mycrl** [comment]
sender:

```log
level=LOG_NOTICE,area=Some("srt::CUDT::postConnect"), message=Some("14:53:55.438000/T11932.N:SRT.cn: @612686939: Connection established to: 192.168.2.88:8080\n")
```

forwarding server:

```log
2024-05-30T06:53:55.458Z INFO  [srt] level=LOG_NOTICE,area=Some("worker_ProcessConnectionRequest"), message=Some("14:53:55.458174/SRT:RcvQ:w1.N:SRT.cn: PASSING request from: 192.168.2.41:57578 to agent:9973767\n")
2024-05-30T06:53:55.459Z INFO  [srt] level=LOG_NOTICE,area=Some("worker_ProcessConnectionRequest"), message=Some("14:53:55.458955/SRT:RcvQ:w1.N:SRT.cn: Listener managed the connection request from: 192.168.2.41:57578 result:waveahand\n")
2024-05-30T06:53:55.461Z INFO  [srt] level=LOG_NOTICE,area=Some("worker_ProcessConnectionRequest"), message=Some("14:53:55.461300/SRT:RcvQ:w1.N:SRT.cn: PASSING request from: 192.168.2.41:57578 to agent:9973767\n")
2024-05-30T06:53:55.461Z INFO  [srt] level=LOG_NOTICE,area=Some("processSrtMsg_HSREQ"), message=Some("14:53:55.461575/SRT:RcvQ:w1.N:SRT.cn: HSREQ/rcv: cmd=1(HSREQ) len=12 vers=0x10503 opts=0xbf delay=30\n")
2024-05-30T06:53:55.463Z INFO  [srt] level=LOG_NOTICE,area=Some("processConnectRequest"), message=Some("14:53:55.463587/SRT:RcvQ:w1.N:SRT.cn: @9973767: listen ret: -1 - conclusion\n")
2024-05-30T06:53:55.463Z INFO  [srt] level=LOG_NOTICE,area=Some("worker_ProcessConnectionRequest"), message=Some("14:53:55.463826/SRT:RcvQ:w1.N:SRT.cn: Listener managed the connection request from: 192.168.2.41:57578 result:waveahand\n")
2024-05-30T06:54:04.855Z INFO  [srt] level=LOG_NOTICE,area=Some("worker_ProcessConnectionRequest"), message=Some("14:54:04.855544/SRT:RcvQ:w1.N:SRT.cn: PASSING request from: 192.168.2.41:43471 to agent:9973767\n")
2024-05-30T06:54:04.855Z INFO  [srt] level=LOG_NOTICE,area=Some("worker_ProcessConnectionRequest"), message=Some("14:54:04.855795/SRT:RcvQ:w1.N:SRT.cn: Listener managed the connection request from: 192.168.2.41:43471 result:waveahand\n")
2024-05-30T06:54:04.860Z INFO  [srt] level=LOG_NOTICE,area=Some("worker_ProcessConnectionRequest"), message=Some("14:54:04.860495/SRT:RcvQ:w1.N:SRT.cn: PASSING request from: 192.168.2.41:43471 to agent:9973767\n")
2024-05-30T06:54:04.860Z INFO  [srt] level=LOG_NOTICE,area=Some("processSrtMsg_HSREQ"), message=Some("14:54:04.860776/SRT:RcvQ:w1.N:SRT.cn: HSREQ/rcv: cmd=1(HSREQ) len=12 vers=0x10503 opts=0xbf delay=30\n")
2024-05-30T06:54:04.862Z INFO  [srt] level=LOG_NOTICE,area=Some("processConnectRequest"), message=Some("14:54:04.862777/SRT:RcvQ:w1.N:SRT.cn: @9973767: listen ret: -1 - conclusion\n")
2024-05-30T06:54:04.863Z INFO  [srt] level=LOG_NOTICE,area=Some("worker_ProcessConnectionRequest"), message=Some("14:54:04.862993/SRT:RcvQ:w1.N:SRT.cn: Listener managed the connection request from: 192.168.2.41:43471 result:waveahand\n")
2024-05-30T06:54:14.578Z INFO  [srt] level=LOG_NOTICE,area=Some("processCtrlAckAck"), message=Some("14:54:14.578040/SRT:RcvQ:w1.N:SRT.in: @9973766: ACKACK out of order, skipping RTT calculation (ACK number: 1123, last ACK sent: 1123, RTT (EWMA): 1397)\n")
2024-05-30T06:55:05.322Z INFO  [srt] level=LOG_WARNING,area=Some("tsbpd"), message=Some("14:55:05.322665/SRT:TsbPd!W:SRT.br: @9973766: RCV-DROPPED 3 packet(s). Packet seqno %803184757 delayed for 0.080 ms\n")
2024-05-30T06:55:15.483Z INFO  [srt] level=LOG_NOTICE,area=Some("processCtrlAckAck"), message=Some("14:55:15.483452/SRT:RcvQ:w1.N:SRT.in: @9973766: ACKACK out of order, skipping RTT calculation (ACK number: 4985, last ACK sent: 4985, RTT (EWMA): 3400)\n")
```

I keep calling srt_send but no data is received and there is no network activity on the sender (except for a small amount of packet reception).

**Comment by ethouris** [comment]
First off: is your application calling `srt_send` with regular time distance between the calls?

Note that in the live mode SRT does not control the input speed, except the case of the bandwidth control. There's an overall high bandwidth limit set designated in the code as `BW_INFINITE`. It can still be set higher than this using the `SRTO_MAXBW` option.
```
const int64_t BW_INFINITE =  1000000000/8;         //Infinite=> 1 Gbps
```

**Comment by mycrl** [comment]
> First off: is your application calling `srt_send` with regular time distance between the calls?
> 
> Note that in the live mode SRT does not control the input speed, except the case of the bandwidth control. There's an overall high bandwidth limit set designated in the code as `BW_INFINITE`. It can still be set higher than this using the `SRTO_MAXBW` option.
> 
> ```
> const int64_t BW_INFINITE =  1000000000/8;         //Infinite=> 1 Gbps
> ```

I didn't deliberately control the sending interval. After reading the h264 packet from the encoder, I split it into appropriate sizes according to the max packet size and sent all the chunks through srt_send.
So my sending rate is related to the encoder output rate. Of course, if the encoder outputs too large nalu, I split it into many chunks, and srt_send will be called a lot in a short time, because I just `split` and then `for { srt_send }`.
Strangely, when the stream is paused (that is, I am still sending, but srt does not help me send it out), srt_send is not affected in any way, nor is it blocked, and I observed the memory usage, and it did not increase (it does not seem to be written to the buffer queue). Srt simply does not send data for me.

**Comment by ethouris** [comment]
And likely that's more-less the problem.

TCP controls the sending speed (and so does SRT in file mode), so you can send all the data at once, it will take care of splitting them into packets and will send them with appropriate speed that could avoid most of the packet loss.

SRT in live mode doesn't do any speed control except for bandwidth limit. If you send all data "at once" (regarding the size limit of the single call and calls in a loop), the underlying data carrying UDP packets will be also sent at once, with only a minimum distance in time that corresponds to the above limit, which is usually much more than your network can carry.

[Here are the details](../docs/features/live-streaming.md).

**Comment by mycrl** [comment]
I have successfully reduced the output bitrate of qsv significantly. This problem has been successfully solved for me, so I will close this issue, but I still hope that srt can pay attention to similar problems.

**Comment by ethouris** [comment]
This works this way in SRT by design. SRT has no possibility to control the sending speed basing on the data timestamps simply because SRT knows nothing about them, nor can it be anyhow informed about them. On this level there's only possible the same thing as TCP does, but this works against the time requirements of the live streaming.

**Comment by mycrl** [comment]
> This works this way in SRT by design. SRT has no possibility to control the sending speed basing on the data timestamps simply because SRT knows nothing about them, nor can it be anyhow informed about them. On this level there's only possible the same thing as TCP does, but this works against the time requirements of the live streaming.

Does this mean that SRT cannot support high bitrate and low latency live broadcast requirements?

The write rate of a high bitrate video stream can never be too low, and backpressure will definitely exist.

**Comment by ethouris** [comment]
Everything is explained in the document I have linked in my previous comment.

There's an internal limit of data rate, which exists because the algorithms for sending originally written in UDT do not work with any infinite limit. But this limit can be still overcome by setting the limit using `SRTO_MAXBW` option.

**Comment by maxsharabayko** [comment]
See also [`SRTO_MAXBW`](https://github.com/Haivision/srt/blob/v1.5.3/docs/API/API-socket-options.md#SRTO_MAXBW).

**Comment by mycrl** [comment]
> See also [`SRTO_MAXBW`](https://github.com/Haivision/srt/blob/v1.5.3/docs/API/API-socket-options.md#SRTO_MAXBW).

@maxsharabayko First of all, thank you for your reply.
My MAXBW has always been -1, the default is 1Gbps (if the documentation is correct)，Obviously my input bit rate is far from 1gbps.
I hate to bother the maintainers by asking stupid questions, but I don't seem to be doing anything wrong at the moment.

**Comment by maxsharabayko** [comment]
@mycrl 
Then it should not be related to `SRTO_MAXBW`.
I read the problem description again.

> The loss is serious under Windows, but I only tested Windows and Android. I think it is related to the high bit rate.

First of all, there were some issues with sending on Windows. In particular, SRT versions before 1.5.3 suffered from issue #973. Essentially, at high bitrates some packets were not sent successfully over UDP by windows socket.
The issue has been resolved in v1.5.3. But a new issue was introduced: sending may be sometimes blocked, even stuck in UDP sending.
The latter has been resolved in PR #2929, to be included in SRT v1.5.4 release. The fix is also available in the latest master.

Please also study the [configuration guideline](https://github.com/Haivision/srt/blob/v1.5.3/docs/API/configuration-guidelines.md) to properly configure the RCV buffer.

To further diagnose your issue, you could take a look into [SRT stats](https://github.com/Haivision/srt/blob/v1.5.3/docs/API/statistics.md). For example, sending may be blocked by the `pktFlowWindow` and `pktFlightSize`. It would be also interesting to learn the amount of retransmissions you have, etc.

**Comment by mycrl** [comment]
> @mycrl Then it should not be related to `SRTO_MAXBW`. I read the problem description again.
> 
> > The loss is serious under Windows, but I only tested Windows and Android. I think it is related to the high bit rate.
> 
> First of all, there were some issues with sending on Windows. In particular, SRT versions before 1.5.3 suffered from issue #973. Essentially, at high bitrates some packets were not sent successfully over UDP by windows socket. The issue has been resolved in v1.5.3. But a new issue was introduced: sending may be sometimes blocked, even stuck in UDP sending. The latter has been resolved in PR #2929, to be included in SRT v1.5.4 release. The fix is also available in the latest master.
> 
> Please also study the [configuration guideline](https://github.com/Haivision/srt/blob/v1.5.3/docs/API/configuration-guidelines.md) to properly configure the RCV buffer.
> 
> To further diagnose your issue, you could take a look into [SRT stats](https://github.com/Haivision/srt/blob/v1.5.3/docs/API/statistics.md). For example, sending may be blocked by the `pktFlowWindow` and `pktFlightSize`. It would be also interesting to learn the amount of retransmissions you have, etc.

Thanks, I'll switch to the master branch and look forward to the 1.5.4 release!
